From 282cc303cd1c0919b45183c1fca3ffa9f5e4f8f7 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Tue, 11 Jul 2023 11:22:49 +0200
Subject: [PATCH 320/841] feat(st-ddr): phy: release upgrade (phyinit)

Upgrade PHY phyinit part to A2022.11 release.
Globally these are minor changes.
Update firmware size information.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git e4e2a7ec9bdfe63cda7c32acaa87d0182fc260ea feat(st-ddr): phy: release upgrade (phyinit)]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I49587349ee718fcaff1caad8083f035ed7463e04
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/323362
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 .../ddr/phy/phyinit/include/ddrphy_phyinit.h  |   2 +-
 .../phyinit/include/ddrphy_phyinit_struct.h   |   3 -
 .../include/ddrphy_phyinit_usercustom.h       |   6 +-
 .../src/ddrphy_phyinit_i_loadpieimage.c       |  20 ++-
 .../src/ddrphy_phyinit_loadpieprodcode.c      | 160 +++++++++---------
 .../src/ddrphy_phyinit_progcsrskiptrain.c     |  49 +++++-
 .../src/ddrphy_phyinit_restore_sequence.c     |   5 -
 .../ddrphy_phyinit_usercustom_saveretregs.c   |   5 -
 8 files changed, 139 insertions(+), 111 deletions(-)

diff --git a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h
index 5f0cc0ebc..b6544ca49 100644
--- a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h
+++ b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit.h
@@ -9,7 +9,7 @@
 
 #include <ddrphy_phyinit_usercustom.h>
 
-#define DDRPHY_PHYINIT_RID 202006
+#define DDRPHY_PHYINIT_RID 20220100
 
 /* Global Structures : instantiated in ddrphy_globals.c */
 extern struct runtime_config runtimeconfig;
diff --git a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h
index 44bc70bd7..d4cf77502 100644
--- a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h
+++ b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h
@@ -950,9 +950,6 @@ struct user_input_advanced {
 					 *     the instance in Acs4AnibDis registers. see register
 					 *     description for details.
 					 *
-					 * \note This feature is BETA and untested. Future PhyInit
-					 *       version will fully enable this feature.
-					 *
 					 * Value | Description
 					 * ----- | ---
 					 *   0x1 | Enable
diff --git a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h
index a70233d94..583faf91d 100644
--- a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h
+++ b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h
@@ -49,13 +49,13 @@
 #endif /* STM32MP_LPDDR4_TYPE */
 
 #if STM32MP_DDR3_TYPE
-#define IMEM_SIZE			0x4b94
+#define IMEM_SIZE			0x4c28
 #define DMEM_SIZE			0x6c8
 #elif STM32MP_DDR4_TYPE
-#define IMEM_SIZE			0x6ac8
+#define IMEM_SIZE			0x6d24
 #define DMEM_SIZE			0x6cc
 #elif STM32MP_LPDDR4_TYPE
-#define IMEM_SIZE			0x7ec6
+#define IMEM_SIZE			0x7e50
 #define DMEM_SIZE			0x67c
 #endif
 #define IMEM_ST_ADDR			0x50000
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c
index 6c2a6e98d..1d0da83cd 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021-2023, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -257,6 +257,14 @@ static void traininghwreg_program(void)
 	regdata = (0xFU << CSR_ACSMCKEENB_LSB);
 	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (C0 | TACSM | CSR_ACSMCTRL13_ADDR)), regdata);
 
+	/*
+	 * - Register: AcsmCtrl1
+	 *   - Fields: AcsmRepCnt
+	 *             Need 19 iterations @ 0.25ui increments to cover 4.5UI
+	 */
+	regdata = (0xEU << CSR_ACSMREPCNT_LSB);
+	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (C0 | TACSM | CSR_ACSMCTRL1_ADDR)), regdata);
+
 	/*
 	 * - Register: TsmByte1, TsmByte2
 	 *   - Dependencies: userinputbasic.numdbyte
@@ -385,6 +393,11 @@ void ddrphy_phyinit_i_loadpieimage(bool skip_training)
 	dfiwrrddatacsconfig_program();
 #endif /* STM32MP_LPDDR4_TYPE */
 
+	/*
+	 * - Register: Seq0BGPR7
+	 *   - Program active CSx for MRS7 during D4 RDIMM frequency change
+	 */
+
 	seq0bdly_program();
 
 	seq0bdisableflag_program(skip_training);
@@ -414,13 +427,8 @@ void ddrphy_phyinit_i_loadpieimage(bool skip_training)
 	 * set for mission mode.  Additionally APB access is Isolated by setting
 	 * MicroContMuxSel.
 	 */
-#if STM32MP_DDR3_TYPE || STM32MP_DDR4_TYPE
 	/* Disabling Ucclk (PMU) and Hclk (training hardware) */
 	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TDRTUB | CSR_UCCLKHCLKENABLES_ADDR)), 0x0U);
-#elif STM32MP_LPDDR4_TYPE
-	/* Disabling Ucclk (PMU) */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TDRTUB | CSR_UCCLKHCLKENABLES_ADDR)), 0x2U);
-#endif /* STM32MP_LPDDR4_TYPE */
 
 	/* Isolate the APB access from the internal CSRs by setting the MicroContMuxSel CSR to 1 */
 	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR)), 0x1U);
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_loadpieprodcode.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_loadpieprodcode.c
index 0ae1fee9a..b6c254260 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_loadpieprodcode.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_loadpieprodcode.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021-2023, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -14,7 +14,7 @@
 #include <platform_def.h>
 
 #if STM32MP_DDR3_TYPE || STM32MP_DDR4_TYPE
-#define PRODCODE_SIZE	171
+#define PRODCODE_SIZE	177
 
 static const uint32_t prodcode_addr[PRODCODE_SIZE] = {
 	0x90000U, 0x90001U, 0x90002U, 0x90003U, 0x90004U, 0x90005U, 0x90029U, 0x9002AU, 0x9002BU,
@@ -35,7 +35,8 @@ static const uint32_t prodcode_addr[PRODCODE_SIZE] = {
 	0x900AAU, 0x900ABU, 0x900ACU, 0x900ADU, 0x900AEU, 0x900AFU, 0x900B0U, 0x900B1U, 0x900B2U,
 	0x900B3U, 0x900B4U, 0x900B5U, 0x900B6U, 0x900B7U, 0x900B8U, 0x900B9U, 0x900BAU, 0x900BBU,
 	0x900BCU, 0x900BDU, 0x900BEU, 0x900BFU, 0x900C0U, 0x900C1U, 0x900C2U, 0x900C3U, 0x900C4U,
-	0x90006U, 0x90007U, 0x90008U, 0x90009U, 0x9000AU, 0x9000BU, 0xD00E7U, 0x90017U, 0x90026U,
+	0x900C5U, 0x900C6U, 0x900C7U, 0x900C8U, 0x900C9U, 0x900CAU, 0x90006U, 0x90007U, 0x90008U,
+	0x90009U, 0x9000AU, 0x9000BU, 0xD00E7U, 0x90017U, 0x90026U,
 };
 
 static const uint16_t prodcode_data[PRODCODE_SIZE] = {
@@ -43,24 +44,24 @@ static const uint16_t prodcode_data[PRODCODE_SIZE] = {
 	0x0448U, 0x0139U, 0x0008U, 0x0478U, 0x0109U, 0x0002U, 0x0010U, 0x0139U, 0x000BU, 0x07C0U,
 	0x0139U, 0x0044U, 0x0633U, 0x0159U, 0x014FU, 0x0630U, 0x0159U, 0x0047U, 0x0633U, 0x0149U,
 	0x004FU, 0x0633U, 0x0179U, 0x0008U, 0x00E0U, 0x0109U, 0x0000U, 0x07C8U, 0x0109U, 0x0000U,
-	0x0001U, 0x0008U, 0x0000U, 0x045AU, 0x0009U, 0x0000U, 0x0448U, 0x0109U, 0x0040U, 0x0633U,
-	0x0179U, 0x0001U, 0x0618U, 0x0109U, 0x40C0U, 0x0633U, 0x0149U, 0x0008U, 0x0004U, 0x0048U,
-	0x4040U, 0x0633U, 0x0149U, 0x0000U, 0x0004U, 0x0048U, 0x0040U, 0x0633U, 0x0149U, 0x0010U,
-	0x0004U, 0x0018U, 0x0000U, 0x0004U, 0x0078U, 0x0549U, 0x0633U, 0x0159U, 0x0D49U, 0x0633U,
-	0x0159U, 0x094AU, 0x0633U, 0x0159U, 0x0441U, 0x0633U, 0x0149U, 0x0042U, 0x0633U, 0x0149U,
-	0x0001U, 0x0633U, 0x0149U, 0x0000U, 0x00E0U, 0x0109U, 0x000AU, 0x0010U, 0x0109U, 0x0009U,
-	0x03C0U, 0x0149U, 0x0009U, 0x03C0U, 0x0159U, 0x0018U, 0x0010U, 0x0109U, 0x0000U, 0x03C0U,
-	0x0109U, 0x0018U, 0x0004U, 0x0048U, 0x0018U, 0x0004U, 0x0058U, 0x000BU, 0x0010U, 0x0109U,
-	0x0001U, 0x0010U, 0x0109U, 0x0005U, 0x07C0U, 0x0109U, 0x0000U, 0x8140U, 0x010CU, 0x0010U,
-	0x8138U, 0x0104U, 0x0008U, 0x0448U, 0x0109U, 0x000FU, 0x07C0U, 0x0109U, 0x0047U, 0x0630U,
-	0x0109U, 0x0008U, 0x0618U, 0x0109U, 0x0008U, 0x00E0U, 0x0109U, 0x0000U, 0x07C8U, 0x0109U,
-	0x0008U, 0x8140U, 0x010CU, 0x0000U, 0x0478U, 0x0109U, 0x0000U, 0x0001U, 0x0008U, 0x0008U,
-	0x0004U, 0x0000U, 0x0008U, 0x07C8U, 0x0109U, 0x0000U, 0x0400U, 0x0106U, 0x0400U, 0x0000U,
-	0x002AU,
+	0x0001U, 0x0008U, 0x0030U, 0x065AU, 0x0009U, 0x0000U, 0x045AU, 0x0009U, 0x0000U, 0x0448U,
+	0x0109U, 0x0040U, 0x0633U, 0x0179U, 0x0001U, 0x0618U, 0x0109U, 0x40C0U, 0x0633U, 0x0149U,
+	0x0008U, 0x0004U, 0x0048U, 0x4040U, 0x0633U, 0x0149U, 0x0000U, 0x0004U, 0x0048U, 0x0040U,
+	0x0633U, 0x0149U, 0x0000U, 0x0658U, 0x0109U, 0x0010U, 0x0004U, 0x0018U, 0x0000U, 0x0004U,
+	0x0078U, 0x0549U, 0x0633U, 0x0159U, 0x0D49U, 0x0633U, 0x0159U, 0x094AU, 0x0633U, 0x0159U,
+	0x0441U, 0x0633U, 0x0149U, 0x0042U, 0x0633U, 0x0149U, 0x0001U, 0x0633U, 0x0149U, 0x0000U,
+	0x00E0U, 0x0109U, 0x000AU, 0x0010U, 0x0109U, 0x0009U, 0x03C0U, 0x0149U, 0x0009U, 0x03C0U,
+	0x0159U, 0x0018U, 0x0010U, 0x0109U, 0x0000U, 0x03C0U, 0x0109U, 0x0018U, 0x0004U, 0x0048U,
+	0x0018U, 0x0004U, 0x0058U, 0x000BU, 0x0010U, 0x0109U, 0x0001U, 0x0010U, 0x0109U, 0x0005U,
+	0x07C0U, 0x0109U, 0x0000U, 0x8140U, 0x010CU, 0x0010U, 0x8138U, 0x0104U, 0x0008U, 0x0448U,
+	0x0109U, 0x000FU, 0x07C0U, 0x0109U, 0x0047U, 0x0630U, 0x0109U, 0x0008U, 0x0618U, 0x0109U,
+	0x0008U, 0x00E0U, 0x0109U, 0x0000U, 0x07C8U, 0x0109U, 0x0008U, 0x8140U, 0x010CU, 0x0000U,
+	0x0478U, 0x0109U, 0x0000U, 0x0001U, 0x0008U, 0x0008U, 0x0004U, 0x0000U, 0x0008U, 0x07C8U,
+	0x0109U, 0x0000U, 0x0400U, 0x0106U, 0x0400U, 0x0000U, 0x002CU,
 	};
 
 #elif STM32MP_LPDDR4_TYPE
-#define PRODCODE_SIZE		478
+#define PRODCODE_SIZE		481
 
 static const uint32_t prodcode_addr[PRODCODE_SIZE] = {
 	0x90000U, 0x90001U, 0x90002U, 0x90003U, 0x90004U, 0x90005U, 0x90029U, 0x9002AU, 0x9002BU,
@@ -77,19 +78,19 @@ static const uint32_t prodcode_addr[PRODCODE_SIZE] = {
 	0x90086U, 0x90087U, 0x90088U, 0x90089U, 0x9008AU, 0x9008BU, 0x9008CU, 0x9008DU, 0x9008EU,
 	0x9008FU, 0x90090U, 0x90091U, 0x90092U, 0x90093U, 0x90094U, 0x90095U, 0x90096U, 0x90097U,
 	0x90098U, 0x90099U, 0x9009AU, 0x9009BU, 0x9009CU, 0x9009DU, 0x9009EU, 0x9009FU, 0x900A0U,
-	0x900A1U, 0x900A2U, 0x900A3U, 0x40000U, 0x40020U, 0x40040U, 0x40060U, 0x40001U, 0x40021U,
-	0x40041U, 0x40061U, 0x40002U, 0x40022U, 0x40042U, 0x40062U, 0x40003U, 0x40023U, 0x40043U,
-	0x40063U, 0x40004U, 0x40024U, 0x40044U, 0x40064U, 0x40005U, 0x40025U, 0x40045U, 0x40065U,
-	0x40006U, 0x40026U, 0x40046U, 0x40066U, 0x40007U, 0x40027U, 0x40047U, 0x40067U, 0x40008U,
-	0x40028U, 0x40048U, 0x40068U, 0x40009U, 0x40029U, 0x40049U, 0x40069U, 0x4000AU, 0x4002AU,
-	0x4004AU, 0x4006AU, 0x4000BU, 0x4002BU, 0x4004BU, 0x4006BU, 0x4000CU, 0x4002CU, 0x4004CU,
-	0x4006CU, 0x4000DU, 0x4002DU, 0x4004DU, 0x4006DU, 0x4000EU, 0x4002EU, 0x4004EU, 0x4006EU,
-	0x4000FU, 0x4002FU, 0x4004FU, 0x4006FU, 0x40010U, 0x40030U, 0x40050U, 0x40070U, 0x40011U,
-	0x40031U, 0x40051U, 0x40071U, 0x40012U, 0x40032U, 0x40052U, 0x40072U, 0x40013U, 0x40033U,
-	0x40053U, 0x40073U, 0x40014U, 0x40034U, 0x40054U, 0x40074U, 0x40015U, 0x40035U, 0x40055U,
-	0x40075U, 0x40016U, 0x40036U, 0x40056U, 0x40076U, 0x40017U, 0x40037U, 0x40057U, 0x40077U,
-	0x40018U, 0x40038U, 0x40058U, 0x40078U, 0x40019U, 0x40039U, 0x40059U, 0x40079U, 0x4001AU,
-	0x4003AU, 0x4005AU, 0x4007AU, 0x900A4U, 0x900A5U, 0x900A6U, 0x900A7U, 0x900A8U, 0x900A9U,
+	0x900A1U, 0x900A2U, 0x900A3U, 0x900A4U, 0x900A5U, 0x900A6U, 0x900A7U, 0x900A8U, 0x900A9U,
+	0x40000U, 0x40020U, 0x40040U, 0x40060U, 0x40001U, 0x40021U, 0x40041U, 0x40061U, 0x40002U,
+	0x40022U, 0x40042U, 0x40062U, 0x40003U, 0x40023U, 0x40043U, 0x40063U, 0x40004U, 0x40024U,
+	0x40044U, 0x40064U, 0x40005U, 0x40025U, 0x40045U, 0x40065U, 0x40006U, 0x40026U, 0x40046U,
+	0x40066U, 0x40007U, 0x40027U, 0x40047U, 0x40067U, 0x40008U, 0x40028U, 0x40048U, 0x40068U,
+	0x40009U, 0x40029U, 0x40049U, 0x40069U, 0x4000AU, 0x4002AU, 0x4004AU, 0x4006AU, 0x4000BU,
+	0x4002BU, 0x4004BU, 0x4006BU, 0x4000CU, 0x4002CU, 0x4004CU, 0x4006CU, 0x4000DU, 0x4002DU,
+	0x4004DU, 0x4006DU, 0x4000EU, 0x4002EU, 0x4004EU, 0x4006EU, 0x4000FU, 0x4002FU, 0x4004FU,
+	0x4006FU, 0x40010U, 0x40030U, 0x40050U, 0x40070U, 0x40011U, 0x40031U, 0x40051U, 0x40071U,
+	0x40012U, 0x40032U, 0x40052U, 0x40072U, 0x40013U, 0x40033U, 0x40053U, 0x40073U, 0x40014U,
+	0x40034U, 0x40054U, 0x40074U, 0x40015U, 0x40035U, 0x40055U, 0x40075U, 0x40016U, 0x40036U,
+	0x40056U, 0x40076U, 0x40017U, 0x40037U, 0x40057U, 0x40077U, 0x40018U, 0x40038U, 0x40058U,
+	0x40078U, 0x40019U, 0x40039U, 0x40059U, 0x40079U, 0x4001AU, 0x4003AU, 0x4005AU, 0x4007AU,
 	0x900AAU, 0x900ABU, 0x900ACU, 0x900ADU, 0x900AEU, 0x900AFU, 0x900B0U, 0x900B1U, 0x900B2U,
 	0x900B3U, 0x900B4U, 0x900B5U, 0x900B6U, 0x900B7U, 0x900B8U, 0x900B9U, 0x900BAU, 0x900BBU,
 	0x900BCU, 0x900BDU, 0x900BEU, 0x900BFU, 0x900C0U, 0x900C1U, 0x900C2U, 0x900C3U, 0x900C4U,
@@ -101,7 +102,7 @@ static const uint32_t prodcode_addr[PRODCODE_SIZE] = {
 	0x900F2U, 0x900F3U, 0x900F4U, 0x900F5U, 0x900F6U, 0x900F7U, 0x900F8U, 0x900F9U, 0x900FAU,
 	0x900FBU, 0x900FCU, 0x900FDU, 0x900FEU, 0x900FFU, 0x90100U, 0x90101U, 0x90102U, 0x90103U,
 	0x90104U, 0x90105U, 0x90106U, 0x90107U, 0x90108U, 0x90109U, 0x9010AU, 0x9010BU, 0x9010CU,
-	0x9010DU, 0x9010EU, 0x9010FU, 0x90110U, 0x90111U, 0x90112U, 0x90113U, 0x90114U,	0x90115U,
+	0x9010DU, 0x9010EU, 0x9010FU, 0x90110U, 0x90111U, 0x90112U, 0x90113U, 0x90114U, 0x90115U,
 	0x90116U, 0x90117U, 0x90118U, 0x90119U, 0x9011AU, 0x9011BU, 0x9011CU, 0x9011DU, 0x9011EU,
 	0x9011FU, 0x90120U, 0x90121U, 0x90122U, 0x90123U, 0x90124U, 0x90125U, 0x90126U, 0x90127U,
 	0x90128U, 0x90129U, 0x9012AU, 0x9012BU, 0x9012CU, 0x9012DU, 0x9012EU, 0x9012FU, 0x90130U,
@@ -114,9 +115,9 @@ static const uint32_t prodcode_addr[PRODCODE_SIZE] = {
 	0x90167U, 0x90168U, 0x90169U, 0x9016AU, 0x9016BU, 0x9016CU, 0x9016DU, 0x9016EU, 0x9016FU,
 	0x90170U, 0x90171U, 0x90172U, 0x90173U, 0x90174U, 0x90175U, 0x90176U, 0x90177U, 0x90178U,
 	0x90179U, 0x9017AU, 0x9017BU, 0x9017CU, 0x9017DU, 0x9017EU, 0x9017FU, 0x90180U, 0x90181U,
-	0x90006U, 0x90007U, 0x90008U, 0x90009U, 0x9000AU, 0x9000BU, 0xD00E7U, 0x90017U, 0x9001FU,
-	0x90026U, 0x400D0U, 0x400D1U, 0x400D2U, 0x400D3U, 0x400D4U, 0x400D5U, 0x400D6U, 0x400D7U,
-	0x2003AU,
+	0x90182U, 0x90183U, 0x90184U, 0x90006U, 0x90007U, 0x90008U, 0x90009U, 0x9000AU, 0x9000BU,
+	0xD00E7U, 0x90017U, 0x9001FU, 0x90026U, 0x400D0U, 0x400D1U, 0x400D2U, 0x400D3U, 0x400D4U,
+	0x400D5U, 0x400D6U, 0x400D7U, 0x2003AU,
 	};
 
 static const uint16_t prodcode_data[PRODCODE_SIZE] = {
@@ -124,50 +125,51 @@ static const uint16_t prodcode_data[PRODCODE_SIZE] = {
 	0x0448U, 0x0139U, 0x0008U, 0x0478U, 0x0109U, 0x0000U, 0x00E8U, 0x0109U, 0x0002U, 0x0010U,
 	0x0139U, 0x000BU, 0x07C0U, 0x0139U, 0x0044U, 0x0633U, 0x0159U, 0x014FU, 0x0630U, 0x0159U,
 	0x0047U, 0x0633U, 0x0149U, 0x004FU, 0x0633U, 0x0179U, 0x0008U, 0x00E0U, 0x0109U, 0x0000U,
-	0x07C8U, 0x0109U, 0x0000U, 0x0001U, 0x0008U, 0x0000U, 0x045AU, 0x0009U, 0x0000U, 0x0448U,
-	0x0109U, 0x0040U, 0x0633U, 0x0179U, 0x0001U, 0x0618U, 0x0109U, 0x40C0U, 0x0633U, 0x0149U,
-	0x0008U, 0x0004U, 0x0048U, 0x4040U, 0x0633U, 0x0149U, 0x0000U, 0x0004U, 0x0048U, 0x0040U,
-	0x0633U, 0x0149U, 0x0010U, 0x0004U, 0x0018U, 0x0000U, 0x0004U, 0x0078U, 0x0549U, 0x0633U,
-	0x0159U, 0x0D49U, 0x0633U, 0x0159U, 0x094AU, 0x0633U, 0x0159U, 0x0441U, 0x0633U, 0x0149U,
-	0x0042U, 0x0633U, 0x0149U, 0x0001U, 0x0633U, 0x0149U, 0x0000U, 0x00E0U, 0x0109U, 0x000AU,
-	0x0010U, 0x0109U, 0x0009U, 0x03C0U, 0x0149U, 0x0009U, 0x03C0U, 0x0159U, 0x0018U, 0x0010U,
-	0x0109U, 0x0000U, 0x03C0U, 0x0109U, 0x0018U, 0x0004U, 0x0048U, 0x0018U, 0x0004U, 0x0058U,
-	0x000BU, 0x0010U, 0x0109U, 0x0001U, 0x0010U, 0x0109U, 0x0005U, 0x07C0U, 0x0109U, 0x0811U,
-	0x0880U, 0x0000U, 0x0000U, 0x4008U, 0x0083U, 0x004FU, 0x0000U, 0x4040U, 0x0083U, 0x0051U,
-	0x0000U, 0x0811U, 0x0880U, 0x0000U, 0x0000U, 0x0720U, 0x000FU, 0x1740U, 0x0000U, 0x0016U,
-	0x0083U, 0x004BU, 0x0000U, 0x0716U, 0x000FU, 0x2001U, 0x0000U, 0x0716U, 0x000FU, 0x2800U,
-	0x0000U, 0x0716U, 0x000FU, 0x0F00U, 0x0000U, 0x0720U, 0x000FU, 0x1400U, 0x0000U, 0x0E08U,
-	0x0C15U, 0x0000U, 0x0000U, 0x0625U, 0x0015U, 0x0000U, 0x0000U, 0x4028U, 0x0080U, 0x0000U,
-	0x0000U, 0x0E08U, 0x0C1AU, 0x0000U, 0x0000U, 0x0625U, 0x001AU, 0x0000U, 0x0000U, 0x4040U,
-	0x0080U, 0x0000U, 0x0000U, 0x2604U, 0x0015U, 0x0000U, 0x0000U, 0x0708U, 0x0005U, 0x0000U,
-	0x2002U, 0x0008U, 0x0080U, 0x0000U, 0x0000U, 0x2604U, 0x001AU, 0x0000U, 0x0000U, 0x0708U,
-	0x000AU, 0x0000U, 0x2002U, 0x4040U, 0x0080U, 0x0000U, 0x0000U, 0x060AU, 0x0015U, 0x1200U,
-	0x0000U, 0x061AU, 0x0015U, 0x1300U, 0x0000U, 0x060AU, 0x001AU, 0x1200U, 0x0000U, 0x0642U,
-	0x001AU, 0x1300U, 0x0000U, 0x4808U, 0x0880U, 0x0000U, 0x0000U, 0x0000U, 0x0790U, 0x011AU,
-	0x0008U, 0x07AAU, 0x002AU, 0x0010U, 0x07B2U, 0x002AU, 0x0000U, 0x07C8U, 0x0109U, 0x0010U,
-	0x0010U, 0x0109U, 0x0010U, 0x02A8U, 0x0129U, 0x0008U, 0x0370U, 0x0129U, 0x000AU, 0x03C8U,
-	0x01A9U, 0x000CU, 0x0408U, 0x0199U, 0x0014U, 0x0790U, 0x011AU, 0x0008U, 0x0004U, 0x0018U,
-	0x000EU, 0x0408U, 0x0199U, 0x0008U, 0x8568U, 0x0108U, 0x0018U, 0x0790U, 0x016AU, 0x0008U,
-	0x01D8U, 0x0169U, 0x0010U, 0x8558U, 0x0168U, 0x0070U, 0x0788U, 0x016AU, 0x1FF8U, 0x85A8U,
-	0x01E8U, 0x0050U, 0x0798U, 0x016AU, 0x0060U, 0x07A0U, 0x016AU, 0x0008U, 0x8310U, 0x0168U,
-	0x0008U, 0xA310U, 0x0168U, 0x000AU, 0x0408U, 0x0169U, 0x006EU, 0x0000U, 0x0068U, 0x0000U,
-	0x0408U, 0x0169U, 0x0000U, 0x8310U, 0x0168U, 0x0000U, 0xA310U, 0x0168U, 0x1FF8U, 0x85A8U,
-	0x01E8U, 0x0068U, 0x0798U, 0x016AU, 0x0078U, 0x07A0U, 0x016AU, 0x0068U, 0x0790U, 0x016AU,
-	0x0008U, 0x8B10U, 0x0168U, 0x0008U, 0xAB10U, 0x0168U, 0x000AU, 0x0408U, 0x0169U, 0x0058U,
-	0x0000U, 0x0068U, 0x0000U, 0x0408U, 0x0169U, 0x0000U, 0x8B10U, 0x0168U, 0x0001U, 0xAB10U,
-	0x0168U, 0x0000U, 0x01D8U, 0x0169U, 0x0080U, 0x0790U, 0x016AU, 0x0018U, 0x07AAU, 0x006AU,
-	0x000AU, 0x0000U, 0x01E9U, 0x0008U, 0x8080U, 0x0108U, 0x000FU, 0x0408U, 0x0169U, 0x000CU,
-	0x0000U, 0x0068U, 0x0009U, 0x0000U, 0x01A9U, 0x0000U, 0x0408U, 0x0169U, 0x0000U, 0x8080U,
-	0x0108U, 0x0008U, 0x07AAU, 0x006AU, 0x0000U, 0x8568U, 0x0108U, 0x00B7U, 0x0790U, 0x016AU,
-	0x001FU, 0x0000U, 0x0068U, 0x0008U, 0x8558U, 0x0168U, 0x000FU, 0x0408U, 0x0169U, 0x000DU,
-	0x0000U, 0x0068U, 0x0000U, 0x0408U, 0x0169U, 0x0000U, 0x8558U, 0x0168U, 0x0008U, 0x03C8U,
-	0x01A9U, 0x0003U, 0x0370U, 0x0129U, 0x0020U, 0x02AAU, 0x0009U, 0x0008U, 0x00E8U, 0x0109U,
-	0x0000U, 0x8140U, 0x010CU, 0x0010U, 0x8138U, 0x0104U, 0x0008U, 0x0448U, 0x0109U, 0x000FU,
-	0x07C0U, 0x0109U, 0x0000U, 0x00E8U, 0x0109U, 0x0047U, 0x0630U, 0x0109U, 0x0008U, 0x0618U,
-	0x0109U, 0x0008U, 0x00E0U, 0x0109U, 0x0000U, 0x07C8U, 0x0109U, 0x0008U, 0x8140U, 0x010CU,
-	0x0000U, 0x0478U, 0x0109U, 0x0000U, 0x0001U, 0x0008U, 0x0008U, 0x0004U, 0x0000U, 0x0008U,
-	0x07C8U, 0x0109U, 0x0000U, 0x0400U, 0x0106U, 0x0400U, 0x0000U, 0x0029U, 0x0068U, 0x0000U,
-	0x0101U, 0x0105U, 0x0107U, 0x010FU, 0x0202U, 0x020aU, 0x020BU, 0x0002U,
+	0x07C8U, 0x0109U, 0x0000U, 0x0001U, 0x0008U, 0x0030U, 0x065AU, 0x0009U, 0x0000U, 0x045AU,
+	0x0009U, 0x0000U, 0x0448U, 0x0109U, 0x0040U, 0x0633U, 0x0179U, 0x0001U, 0x0618U, 0x0109U,
+	0x40C0U, 0x0633U, 0x0149U, 0x0008U, 0x0004U, 0x0048U, 0x4040U, 0x0633U, 0x0149U, 0x0000U,
+	0x0004U, 0x0048U, 0x0040U, 0x0633U, 0x0149U, 0x0000U, 0x0658U, 0x0109U, 0x0010U, 0x0004U,
+	0x0018U, 0x0000U, 0x0004U, 0x0078U, 0x0549U, 0x0633U, 0x0159U, 0x0D49U, 0x0633U, 0x0159U,
+	0x094AU, 0x0633U, 0x0159U, 0x0441U, 0x0633U, 0x0149U, 0x0042U, 0x0633U, 0x0149U, 0x0001U,
+	0x0633U, 0x0149U, 0x0000U, 0x00E0U, 0x0109U, 0x000AU, 0x0010U, 0x0109U, 0x0009U, 0x03C0U,
+	0x0149U, 0x0009U, 0x03C0U, 0x0159U, 0x0018U, 0x0010U, 0x0109U, 0x0000U, 0x03C0U, 0x0109U,
+	0x0018U, 0x0004U, 0x0048U, 0x0018U, 0x0004U, 0x0058U, 0x000BU, 0x0010U, 0x0109U, 0x0001U,
+	0x0010U, 0x0109U, 0x0005U, 0x07C0U, 0x0109U, 0x0811U, 0x0880U, 0x0000U, 0x0000U, 0x4008U,
+	0x0083U, 0x004FU, 0x0000U, 0x4040U, 0x0083U, 0x0051U, 0x0000U, 0x0811U, 0x0880U, 0x0000U,
+	0x0000U, 0x0720U, 0x000FU, 0x1740U, 0x0000U, 0x0016U, 0x0083U, 0x004BU, 0x0000U, 0x0716U,
+	0x000FU, 0x2001U, 0x0000U, 0x0716U, 0x000FU, 0x2800U, 0x0000U, 0x0716U, 0x000FU, 0x0F00U,
+	0x0000U, 0x0720U, 0x000FU, 0x1400U, 0x0000U, 0x0E08U, 0x0C15U, 0x0000U, 0x0000U, 0x0625U,
+	0x0015U, 0x0000U, 0x0000U, 0x4028U, 0x0080U, 0x0000U, 0x0000U, 0x0E08U, 0x0C1AU, 0x0000U,
+	0x0000U, 0x0625U, 0x001AU, 0x0000U, 0x0000U, 0x4040U, 0x0080U, 0x0000U, 0x0000U, 0x2604U,
+	0x0015U, 0x0000U, 0x0000U, 0x0708U, 0x0005U, 0x0000U, 0x2002U, 0x0008U, 0x0080U, 0x0000U,
+	0x0000U, 0x2604U, 0x001AU, 0x0000U, 0x0000U, 0x0708U, 0x000AU, 0x0000U, 0x2002U, 0x4040U,
+	0x0080U, 0x0000U, 0x0000U, 0x060AU, 0x0015U, 0x1200U, 0x0000U, 0x061AU, 0x0015U, 0x1300U,
+	0x0000U, 0x060AU, 0x001AU, 0x1200U, 0x0000U, 0x0642U, 0x001AU, 0x1300U, 0x0000U, 0x4808U,
+	0x0880U, 0x0000U, 0x0000U, 0x0000U, 0x0790U, 0x011AU, 0x0008U, 0x07AAU, 0x002AU, 0x0010U,
+	0x07B2U, 0x002AU, 0x0000U, 0x07C8U, 0x0109U, 0x0010U, 0x0010U, 0x0109U, 0x0010U, 0x02A8U,
+	0x0129U, 0x0008U, 0x0370U, 0x0129U, 0x000AU, 0x03C8U, 0x01A9U, 0x000CU, 0x0408U, 0x0199U,
+	0x0014U, 0x0790U, 0x011AU, 0x0008U, 0x0004U, 0x0018U, 0x000EU, 0x0408U, 0x0199U, 0x0008U,
+	0x8568U, 0x0108U, 0x0018U, 0x0790U, 0x016AU, 0x0008U, 0x01D8U, 0x0169U, 0x0010U, 0x8558U,
+	0x0168U, 0x1FF8U, 0x85A8U, 0x01E8U, 0x0050U, 0x0798U, 0x016AU, 0x0060U, 0x07A0U, 0x016AU,
+	0x0008U, 0x8310U, 0x0168U, 0x0008U, 0xA310U, 0x0168U, 0x000AU, 0x0408U, 0x0169U, 0x006EU,
+	0x0000U, 0x0068U, 0x0000U, 0x0408U, 0x0169U, 0x0000U, 0x8310U, 0x0168U, 0x0000U, 0xA310U,
+	0x0168U, 0x1FF8U, 0x85A8U, 0x01E8U, 0x0068U, 0x0798U, 0x016AU, 0x0078U, 0x07A0U, 0x016AU,
+	0x0068U, 0x0790U, 0x016AU, 0x0008U, 0x8B10U, 0x0168U, 0x0008U, 0xAB10U, 0x0168U, 0x000AU,
+	0x0408U, 0x0169U, 0x0058U, 0x0000U, 0x0068U, 0x0000U, 0x0408U, 0x0169U, 0x0000U, 0x8B10U,
+	0x0168U, 0x0001U, 0xAB10U, 0x0168U, 0x0000U, 0x01D8U, 0x0169U, 0x0080U, 0x0790U, 0x016AU,
+	0x0018U, 0x07AAU, 0x006AU, 0x000AU, 0x0000U, 0x01E9U, 0x0008U, 0x8080U, 0x0108U, 0x000FU,
+	0x0408U, 0x0169U, 0x000CU, 0x0000U, 0x0068U, 0x0009U, 0x0000U, 0x01A9U, 0x0000U, 0x0408U,
+	0x0169U, 0x0000U, 0x8080U, 0x0108U, 0x0008U, 0x07AAU, 0x006AU, 0x0000U, 0x8568U, 0x0108U,
+	0x00B7U, 0x0790U, 0x016AU, 0x001FU, 0x0000U, 0x0068U, 0x0008U, 0x8558U, 0x0168U, 0x000FU,
+	0x0408U, 0x0169U, 0x000DU, 0x0000U, 0x0068U, 0x0000U, 0x0408U, 0x0169U, 0x0000U, 0x8558U,
+	0x0168U, 0x0008U, 0x03C8U, 0x01A9U, 0x0003U, 0x0370U, 0x0129U, 0x0020U, 0x02AAU, 0x0009U,
+	0x0008U, 0x00E8U, 0x0109U, 0x0000U, 0x8140U, 0x010CU, 0x0010U, 0x8138U, 0x0104U, 0x0008U,
+	0x0448U, 0x0109U, 0x000FU, 0x07C0U, 0x0109U, 0x0000U, 0x00E8U, 0x0109U, 0x0047U, 0x0630U,
+	0x0109U, 0x0008U, 0x0618U, 0x0109U, 0x0008U, 0x00E0U, 0x0109U, 0x0000U, 0x07C8U, 0x0109U,
+	0x0008U, 0x8140U, 0x010CU, 0x0000U, 0x0478U, 0x0109U, 0x0000U, 0x0001U, 0x0008U, 0x0008U,
+	0x0004U, 0x0000U, 0x0008U, 0x07C8U, 0x0109U, 0x0000U, 0x0400U, 0x0106U, 0x0400U, 0x0000U,
+	0x002BU, 0x0069U, 0x0000U, 0x0101U, 0x0105U, 0x0107U, 0x010FU, 0x0202U, 0x020AU, 0x020BU,
+	0x0002U,
 	};
 #endif /* STM32MP_LPDDR4_TYPE */
 
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_progcsrskiptrain.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_progcsrskiptrain.c
index af01fe1df..9dcf8c610 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_progcsrskiptrain.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_progcsrskiptrain.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021-2023, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -347,9 +347,9 @@ static void rxendly_program(bool skip_train)
 		/* Compensate for pptenrxenbackoff */
 		zerobackoff_x1000 = (1000 * 24) / 32;
 		if (userinputadvanced.lp4rxpreamblemode[pstate] == 1) {
-			backoff_x1000 = (1000 * 1) - ((1000 * 4) / 32);
+			backoff_x1000 = (1000 * 1) - ((1000 * 2) / 32);
 		} else {
-			backoff_x1000 = (1000 * userinputadvanced.rxenbackoff) - ((1000 * 4) / 32);
+			backoff_x1000 = (1000 * userinputadvanced.rxenbackoff) - ((1000 * 2) / 32);
 		}
 
 		if (skip_train && !userinputadvanced.disableretraining) {
@@ -914,15 +914,46 @@ static void acsmctrl23_program(void)
 static void pllforcecal_plldacvalin_program(void)
 {
 	int dacval_in = 0x10;
-	int force_cal = 1;
+	int force_cal = 0x1;
+	int pllencal = 0x1;
 	int maxrange = 0x1f;
-	uint16_t pllctrl3;
+	int pstate;
+	uint16_t pllctrl3_gpr;
+	uint16_t pllctrl3_startup;
 
-	pllctrl3 = (uint16_t)((dacval_in << CSR_PLLDACVALIN_LSB) |
-			      (force_cal << CSR_PLLFORCECAL_LSB) |
-			      (maxrange << CSR_PLLMAXRANGE_LSB));
+	for (pstate = 0; pstate < userinputbasic.numpstates; pstate++) {
+		int p_addr = pstate << 20;
 
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TMASTER | CSR_PLLCTRL3_ADDR)), pllctrl3);
+		/* Just to test */
+		switch (pstate) {
+		case 0:
+			dacval_in = 0x10;
+			break;
+		case 1:
+			dacval_in = 0x11;
+			break;
+		case 2:
+			dacval_in = 0x12;
+			break;
+		case 3:
+			dacval_in = 0x13;
+			break;
+		default:
+			dacval_in = 0x14;
+			break;
+		}
+
+		pllctrl3_startup = (uint16_t)((dacval_in << CSR_PLLDACVALIN_LSB) |
+					      (maxrange << CSR_PLLMAXRANGE_LSB));
+		pllctrl3_gpr = pllctrl3_startup |
+			       (uint16_t)((force_cal << CSR_PLLFORCECAL_LSB) |
+					  (pllencal << CSR_PLLENCAL_LSB));
+
+		mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TMASTER | CSR_PLLCTRL3_ADDR)),
+			      pllctrl3_startup);
+		mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (p_addr | TINITENG |
+							      CSR_SEQ0BGPR6_ADDR)), pllctrl3_gpr);
+	}
 }
 
 /*
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_restore_sequence.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_restore_sequence.c
index 108c8ebe5..54c3d379c 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_restore_sequence.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_restore_sequence.c
@@ -66,13 +66,8 @@ int ddrphy_phyinit_restore_sequence(void)
 	/*
 	 * -# Write the UcclkHclkEnables CSR to disable the appropriate clocks after all reads done.
 	 */
-#if STM32MP_DDR3_TYPE || STM32MP_DDR4_TYPE
 	/* Disabling Ucclk (PMU) and Hclk (training hardware) */
 	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TDRTUB | CSR_UCCLKHCLKENABLES_ADDR)), 0x0U);
-#elif STM32MP_LPDDR4_TYPE
-	/* Disabling Ucclk (PMU) */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TDRTUB | CSR_UCCLKHCLKENABLES_ADDR)), 0x2U);
-#endif /* STM32MP_LPDDR4_TYPE */
 
 	/*
 	 * -# Write the MicroContMuxSel CSR to 0x1 to isolate the internal CSRs during mission mode.
diff --git a/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c b/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
index df19299cd..ecb08c17d 100644
--- a/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
+++ b/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
@@ -403,13 +403,8 @@ int ddrphy_phyinit_usercustom_saveretregs(void)
 	 * --------------------------------------------------------------------------
 	 */
 
-#if STM32MP_DDR4_TYPE
 	/* Disabling Ucclk (PMU) and Hclk (training hardware) */
 	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TDRTUB | CSR_UCCLKHCLKENABLES_ADDR)), 0x0U);
-#elif STM32MP_LPDDR4_TYPE
-	/* Disabling Ucclk (PMU) */
-	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TDRTUB | CSR_UCCLKHCLKENABLES_ADDR)), 0x2U);
-#endif /* STM32MP_LPDDR4_TYPE */
 
 	mmio_write_16((uintptr_t)(DDRPHYC_BASE + 4 * (TAPBONLY | CSR_MICROCONTMUXSEL_ADDR)), 0x1U);
 
-- 
2.47.2

